File: xuri-rpc-test/testscript/base/ws_rpc.py
import asyncio
import json
import websockets
from xuri_rpc import PlainProxyManager, RunnableProxyManager, MessageReceiver, Client, asProxy, getMessageReceiver, setHostId,dict2obj

from util import assertV
port=18081
#无限可能？
async def getClient():

    setHostId('backend')
    from xuri_rpc import setDebugFlag
    setDebugFlag(True)


    class Sender:
        def __init__(self, ws):
            self.ws = ws

        async def send(self, message):
            # message is an object can be jsonified
            await self.ws.send(json.dumps(message))

    setHostId('frontend')
    client = Client()

    ws = await websockets.connect(f'ws://localhost:{port}')

    async def on_message(data):
        await getMessageReceiver().onReceiveMessage(json.loads(data), client)
        print(f'收到服务器消息: {data}')

    # Run message reception in the background
    async def listen():
        async for message in ws:
            asyncio.ensure_future(on_message(message))

    # Start listening in the background
    asyncio.ensure_future(listen())

    client.setSender(Sender(ws))

    return client
def onServe( mainObject ):

    # 设置hostName
    setHostId('backend')

    # 创建一个Sender
    class Sender:
        def __init__(self, ws):
            self.ws = ws

        async def send(self, message):
            await self.ws.send(json.dumps(message))

    # 设置用于提供起始方法的main对象
    getMessageReceiver().setMain(dict2obj(mainObject))

    async def handle_connection(websocket, path):
        # 创建一个client用于发送返回信息
        client = Client()
        client.setSender(Sender(websocket))

        async for data in websocket:
            try:
                # 处理接收到的信息
                asyncio.ensure_future(getMessageReceiver().onReceiveMessage(json.loads(data), client))
            except Exception as e:
                print('客户端连接错误:', e)

    async def main():
        server = await websockets.serve(handle_connection, "localhost", port)
        await server.wait_closed()
    return main()

File: xuri-rpc-test/testscript/base/ws-rpc.js
import { PlainProxyManager,RunnableProxyManager,MessageReceiver,Client,asProxy,getMessageReceiver,setHostId, setDebugFlag } from '../../src/index.js'
import {WebSocket} from 'ws'
import { assertV } from './util.js'

import { WebSocketServer } from 'ws'

const port=18081

export const getClient=async ()=>{
    
    //define a sender
    class Sender {
        constructor(ws) {
            this.ws = ws
        }
        async send(message) {
            //message is an object can be jsonified
            this.ws.send(JSON.stringify(message))
        }
    }
    setHostId('frontend')
    let client = new Client()

    const ws = new WebSocket('ws://localhost:' + port);
    await new Promise((resolve, reject) => {
        ws.on('open', resolve);
        ws.on('error', (e) => {
            console.error(e)
            assertV(false, 'connect error')
        });
    });
    ws.on('message', (data) => {
        getMessageReceiver().onReceiveMessage(JSON.parse(data), client)
        console.log(`收到服务器消息: ${data}`);
    })
    client.setSender(new Sender(ws))

    return client
}

export const onServe=(mainObject)=>{
    // 设置hostName
    setHostId('backend')
    //创建一个Sender
    class Sender{
        constructor(ws){
            this.ws=ws
        }
        async send(message){
            this.ws.send(JSON.stringify(message))
        }
    }
    //设置用于提供起始方法的main对象
    getMessageReceiver().setMain(mainObject)
    
    
    const wss = new WebSocketServer({ port: port });
    wss.on('connection', (ws, request) => {
      
      //创建一个client用于发送返回信息
      let client=new Client()
      client.setSender(new Sender(ws))
    
      ws.on('message', (data) => {
    
        //here handle message
        //处理接收到的信息
        getMessageReceiver().onReceiveMessage(JSON.parse(data),client)
    
      });
    
      ws.on('error', (error) => {
        console.error('客户端连接错误:', error);
        assertV(false, 'connect error')
      });
    });
    wss.on('error', (error) => {
      console.error('服务器错误:', error);
      assertV(false, 'connect error')
    });
}